rfcs/20200520-tensor-float-32.md

```diff
@@ -0,0 +1,80 @@
+# TensorFloat-32 in TensorFlow
+
+| Status        | Proposed                                             |
+:-------------- |:---------------------------------------------------- |
+| **RFC #**     | [247](https://github.com/tensorflow/community/pull/247) |
+| **Author(s)** | Reed Wanderman-Milne (reedwm@google.com)             |
+| **Sponsor**   | Sanjoy Das (sanjoy@google.com)                 |
+| **Updated**   | 2020-05-20                                           |
+
+## Objective
+
+Allow [TensorFloat-32](https://blogs.nvidia.com/blog/2020/05/14/tensorfloat-32-precision-format) to be used in TensorFlow to improve performance.
+
+## Motivation
+
+[NVIDIA Ampere](https://www.nvidia.com/en-us/data-center/nvidia-ampere-gpu-architecture/), an upcoming generation of NVIDIA GPUs announced at GTC 2020, introduces a new numeric format called TensorFloat-32, or TF32 for short.
+TF32 has the range of float32/bfloat16 (i.e. 8 bits of exponent) and the precision of fp16 (i.e. 10 bits of mantissa).
+It is not an in-memory format, but tensor cores natively support it as a computation format.
+TF32 should not be thought of as an in-memory dtype but instead a computation mode that increases performance and decreases numeric precision for certain float32 operations.
+NVIDIA has not found any cases where TF32 reduces the convergence of deep learning models.
+
+Upcoming versions of cuDNN, cuBLAS, and other CUDA libraries will expose a mode of execution that has float32 inputs and outputs, but internally truncates float32 to TF32 and uses tensor cores.  This is expected to be sufficiently accurate to reach the same convergence as the “full” float32 mode of execution but significantly faster.  Each element still takes four bytes, so there is still a memory and performance penalty compared to using float16 or bfloat16.
+
+As TF32 is only usable by tensor cores, it can only be used for matrix multiplications and other ops implemented in terms of matrix multiplications, such as convolutions. It is not used for pointwise ops or reductions.
+
+TF32 will benefit users who run float32 models on Ampere GPUs, so we need an API to allow these users to enable TF32. 
+
+## Design Proposal
+
+In TensorFlow, TF32 can be enabled for supported ops on Ampere GPUs with the following call:
+
+```python
+tf.config.allow_tensor_float_32_execution(True)
+```
+
+The word "allow" emphasizes only certain devices (Ampere GPUs) and ops (such as matmuls and convolutions) will be affected. Once enabled, all local and remote Ampere GPUs use TF32 for supported float32 ops.
+
+Passing `False` to `allow_tensor_float_32_execution` will disable TF32 if already enabled.
```

Comment:
What are the use-cases for toggling between these? Are there potentially issues with moving between the two in a single program?

Reply:
Does the UpdateContext RPC support this particular use case? I imagine one could use the `cluster_device_attributes` field to pass on the TF32 mode toggle, but it looks like it will update the remote session when `cluster_device_attributes` is not empty in current codebase. Not sure if setting TF32 worth carrying out such a heavy(?) operation.

Reply:
Right now UpdateContext is only  used to add/remove machines/devices. But we can add a field to allow it to also update whether TF32 is used. We can altneratively add an option to [QueueItem](https://github.com/tensorflow/tensorflow/blob/6f22fa93766d18f45e2baac140bded0bc3e5e681/tensorflow/core/protobuf/eager_service.proto#L57) in case there are ordering concerns between enabling/disabling TF32 and executing an op.

I asked internally, and `cluster_device_attributes` is only useful for propogating device information to other machines in the cluster. It is not intended for use in communicating information about how ops should run, but only fundamental information about the devices themselves.

See [this post](https://github.com/tensorflow/tensorflow/pull/39764#issuecomment-635055025) for some more context.

> Not sure if setting TF32 worth carrying out such a heavy(?) operation.

Yeah, this will be a heavy operation. But users should only set/unset TF32 very rarely: only at the beginning of the model and between running one model and the next (or one test and the next). So I think it's OK.

I updated the RFC based on this discussion. I added a paragraph to the "Remote Devices" section and added a new paragraph to "Alternatives considered"